### PR TITLE
solana-ledger-tool command to drop column family for downgrade scenarios

### PR DESCRIPTION
**Pull request by jbiseda**
#### Problem

Suppose a new validator version adds a new column family to rocksdb. In order to be able to open rocksdb with a downgraded validator version (that is unaware of the new column), the new CF must be dropped. We typically try to backport columns "stubs" one version back, but this command is still useful in instances such as wanting to downgrade to a version prior to the stub backport. rocksdb `ldb` tool may not be available to perform this operation.

#### Summary of Changes

add new `solana-ledger-tool` command to drop a column family from rocksdb.

Fixes #

### CODE REVIEW

File: ledger-tool/src/main.rs [hunk 1]
```diff
@@ -1765,6 +1765,20 @@ fn main() {
                         .help("Number of slots in the output"),
                 )
         )
+        .subcommand(
+            SubCommand::with_name("drop-cf")
+                .setting(AppSettings::Hidden)
```
Comment (jbiseda) [inline ledger-tool/src/main.rs:1770]: Thoughts on whether this should be hidden?

**Review by steviez — COMMENTED** [review]
On the surface, things look good right now. However, I'll finish my review later, need to look under the hood a little bit

Additionally, added @yhchiang-sol for review as our resident RocksDB expert, and this potentially affecting something he is working on

**Comment by jbiseda** [comment]
> My initial thinking is hidden. This command has the potential to render a rocksdb database unreadable if used incorrectly.

An additional safety check would be to require a `--force` flag in order to actually perform the operation similar to `git clean`. This could be done instead of or in addition to leaving it hidden.

> If things really did hit the fan and we needed to do a major version rollback where the old version didn't have some column, we could communicate instructions on using this command at that point

Yeah, my thinking was that this would be for emergency use. If someone is scrambling to recover by downgrading this would be available rather than providing instructions to build rocksdb tools and use `ldb`.

**Comment by t-nelson** [comment]
do we really want this?  seems like the kind of thing that gets added then used as an excuse to do the wrong thing in the future

**Comment by jbiseda** [comment]
> do we really want this? seems like the kind of thing that gets added then used as an excuse to do the wrong thing in the future

I'm also fine with the answer being to use `ldb`. My thinking was that some people would delete the ledger and start over if they have to build/install a separate tool. Then again, they may just take that action anyway.

**Comment by steviez** [comment]
> do we really want this? seems like the kind of thing that gets added then used as an excuse to do the wrong thing in the future

A few times I've had to bisect pretty far back. As I would shift back and forward across the the commit that added new column, I would cherry-pick the stub column so I could open the ledger. This could have alleviated some merge concerns in that process.

I agree that use of the command should be pretty limited, and moreso, should primarily be by developers for debug instances when necessary (as opposed to by node operators). But, we make mistakes / stuff happens, so I don't think having another tool available is the worst thing.

**Comment by yhchiang-sol** [comment]
If it's for backward compatibility, the rocksdb-options-file approach (described in https://github.com/solana-labs/solana/issues/23665) might be a better solution for longer-term.  Everytime when a rocksdb instance is successfully opened, rocksdb will persist the set of options for all column families that are used to open the instance to a file, and we can use `load_latest_options` (which I am currently porting in rocksdb C API in https://github.com/facebook/rocksdb/pull/10152) to construct the rocksdb options on-the-fly when we open our ledger store. 

But in case a user roll-back to a version where the on-the-fly rocksdb options construction is not available, then downloading the latest ledger-tool and removing the column family might be a faster way.

File: ledger-tool/src/main.rs (hunk 1 above)
Comment (yhchiang-sol) [reply ledger-tool/src/main.rs:1770]: Making it hidden looks good to me as this command is only used during major version rollback.  In addition, incorrectly using this command could make the ledger store unreadable or inconsistent.

**Comment by mvines** [comment]
> rocksdb `ldb` tool may not be available to perform this operation.

What is meant by this? If there's an existing tool that can be simply installed on the machine easily that'll drop columns, documenting that process seems superior to writing more code that we need to maintain

**Comment by yhchiang-sol** [comment]
> rocksdb ldb tool may not be available to perform this operation.

> What is meant by this? If there's an existing tool that can be simply installed on the machine easily that'll drop columns, documenting that process seems superior to writing more code that we need to maintain

[ldb](https://github.com/facebook/rocksdb/wiki/Administration-and-Data-Access-Tool) is the db tool offered by rocksdb (ldb in rocksdb is similar to ledger-tool in solana).  It has several commands including the [drop column family command](https://github.com/facebook/rocksdb/blob/main/tools/ldb_cmd.cc#L1657).

**Comment by t-nelson** [comment]
assuming `ldb` is readily available on linux and macos, i'd take the docs route

**Comment by steviez** [comment]
> assuming `ldb` is readily available on linux and macos, i'd take the docs route

https://formulae.brew.sh/formula/rocksdb
https://packages.debian.org/sid/rocksdb-tools

**Comment by jbiseda** [comment]
I'll add this to docs.

**Comment by jbiseda** [comment]
document using ldb to drop rocksdb column families for downgrade scenarios #26284

**Pull request closed without merging**